internal/disk/disk.go

```diff
@@ -1,228 +1,148 @@
-			rootIdx = idx
-		}
-	}
-	return rootIdx
+// A VolumeContainer is a container that is able to create new volumes.
+//
+// CreateVolume creates a new volume with the given mount point and
+// size and returns it.
+type VolumeContainer interface {
+	CreateVolume(mountpoint string, size uint64) (Entity, error)
 }
 
-// dynamically calculate and update the start point
-// for each of the existing partitions
-// return the updated start point
-func (pt *PartitionTable) updatePartitionStartPointOffsets(start uint64) uint64 {
-	var rootIdx = -1
-	for i := range pt.Partitions {
-		partition := &pt.Partitions[i]
-		if partition.Filesystem != nil && partition.Filesystem.Mountpoint == "/" {
-			rootIdx = i
-			continue
-		}
-		partition.Start = start
-		start += partition.Size
-	}
-	pt.Partitions[rootIdx].Start = start
-	return start
+// A UniqueEntity is an entity that can be uniquely identified via a UUID.
+//
+// GenUUID generates a UUID for the entity if it does not yet have one.
+type UniqueEntity interface {
+	Entity
+	GenUUID(rng *rand.Rand)
 }
 
-func (pt *PartitionTable) updateRootPartition(rootPartition Partition) {
-	pt.Partitions[pt.RootPartitionIndex()] = rootPartition
+// FSSpec for a filesystem (UUID and Label); the first field of fstab(5)
+type FSSpec struct {
+	UUID  string
+	Label string
 }
 
-func (pt *PartitionTable) getPartitionTableSize() uint64 {
-	var size uint64
-	for _, p := range pt.Partitions {
-		size += p.Size
-	}
-	return size
+type FSTabOptions struct {
+	// The fourth field of fstab(5); fs_mntops
+	MntOps string
+	// The fifth field of fstab(5); fs_freq
+	Freq uint64
+	// The sixth field of fstab(5); fs_passno
+	PassNo uint64
 }
 
-// Converts Partition to osbuild.QEMUPartition that encodes the same partition.
-func (p Partition) QEMUPartition() osbuild.QEMUPartition {
-	var fs *osbuild.QEMUFilesystem
-	if p.Filesystem != nil {
-		f := p.Filesystem.QEMUFilesystem()
-		fs = &f
-	}
-	return osbuild.QEMUPartition{
-		Start:      p.Start,
-		Size:       p.Size,
-		Type:       p.Type,
-		Bootable:   p.Bootable,
-		UUID:       p.UUID,
-		Filesystem: fs,
+// uuid generator helpers
+
+// GeneratesnewRandomUUIDFromReader generates a new random UUID (version
+// 4 using) via the given random number generator.
+func newRandomUUIDFromReader(r io.Reader) (uuid.UUID, error) {
+	var id uuid.UUID
+	_, err := io.ReadFull(r, id[:])
+	if err != nil {
+		return uuid.Nil, err
 	}
+	id[6] = (id[6] & 0x0f) | 0x40 // Version 4
+	id[8] = (id[8] & 0x3f) | 0x80 // Variant is 10
+	return id, nil
 }
 
-// Converts Filesystem to osbuild.QEMUFilesystem that encodes the same fs.
-func (fs Filesystem) QEMUFilesystem() osbuild.QEMUFilesystem {
-	return osbuild.QEMUFilesystem{
-		Type:       fs.Type,
-		UUID:       fs.UUID,
-		Label:      fs.Label,
-		Mountpoint: fs.Mountpoint,
-	}
+// NewRandomVolIDFromReader creates a random 32 bit hex string to use as a
+// volume ID for FAT filesystems
+func NewRandomVolIDFromReader(r io.Reader) (string, error) {
```

Comment:
I wonder whether this function really needs to be exported. The function to generate UUID is not exported and the fact that some filesystems use VolumeID and some UUID could be handled by `Filesystem.GenUUID()` based on the FS type. Especially when it is used only in a single place in `internal/distro/rhel90/pipelines.go`.

Reply:
Good point! I think I like the idea of just doing it in the filesystem itself. I do wonder if it can wait to be done in a follow up though.